### DISCUSSION TITLE
Impersonate Feature in RC Branch needed help.

### DISCUSSION THREAD

**Discussion by LeCyreaxYT** [pocketbase/pocketbase#5915] [Q&A] [answered]
Hey, 

I'm currently working on a system where I can impersonate my user's for Testing and get his View. Problem now is when i authStore.save(token, null) and then I execute authRefresh and got this error "The current auth token is not refreshable" is there a way to impersonate a user with an authRefresh so can pull my current user Data to Store it in localStorage or something. 

I hope there is a sulution.

My Code: 
```typescript
<script lang="ts">
export default defineComponent({
    name: "ModalImpersonate",
    emits: ['close'],
    setup(props) {
        const nuxtApp = useNuxtApp();
        const token = ref('');
        const {notify} = useNotification();

        const submitToken = async () => {
            nuxtApp.$pocketbase.getDb().authStore.save(token.value, null);
            try {
              await nuxtApp.$pocketbase.getCurrentUser()
              notify({
                group: 'notify',
                type: 'success',
                title: 'Erfolg',
                text: 'Sie sind jetzt als ein anderer Benutzer angemeldet.',
              });
            } catch (e) {
              notify({
                group: 'notify',
                type: 'error',
                title: 'Fehler',
                text: 'Der Token ist ungültig.',
              });
            }

        }

        return { props, submitToken, token }
    },
    props: {
        show: {
            type: Boolean,
            required: true
        },
    },
})
</script>
```

The function getCurrentUser execute this.
```typescript
try {
   this.db.authStore.isValid && await this.db.collection('users').authRefresh({
      expand: "unit,rank"
   });
} catch (e) {
   this.db.authStore.clear(); // after this call pb.authStore.isValid will be false
}
```

**Comment by chnfyi** [comment]
I think that impersonate tokens are non-refreshable. You can try to send a request as the impersonated user or maybe to repeat an impersonation for the same user to retrieve data and store it.

**Reply by LeCyreaxYT** [reply]
Yeah, i think this is mean with this. But i want to simulate a view of this user, so i want to make a refresh able.

**Reply by ganigeorgiev** [reply, marked as answer]
@LeCyreaxYT The above answer is correct and there is no option out of the box to make it refreshable.

Depending on your use case, one possible workaround could be to read the JWT payload and call `getOne` instead of `authRefresh` if the token is from impersonate action, aka. something like (_I haven't tested it_):

```js
import PocketBase, { getTokenPayload } from "pocketbase/pocketbase"

...

if (this.db.authStore.isValid) {
    try {
        let payload = getTokenPayload(this.db.authStore.token);

        if (payload.refreshable) {
            await this.db.collection('users').authRefresh({ expand: "unit,rank" });
        } else {
            this.db.authStore.save(
                this.db.authStore.token,
                await this.db.collection('users').getOne(payload.id, { expand: "unit,rank" }),
            )
        }
    } catch (e) {
       this.db.authStore.clear();
    }
} else {
   this.db.authStore.clear();
}
```

_Side-note: since it is not clear how `this.db` was initialized, if the above is used in a SSR context make sure to check "Problem 1" from https://github.com/pocketbase/pocketbase/discussions/5313._

**Reply by einarpersson** [reply]
> "I think that impersonate tokens are non-refreshable"

This should be added to the docs [here](https://pocketbase.io/docs/authentication/#users-impersonation) (I couldnt get it to work and now I understand why). I thought the returned client would be identical as I saw no note saying anything else.

**Reply by ganigeorgiev** [reply]
@einarpersson It is already mentioned:
> The generated impersonate auth tokens can have custom duration but are not refreshable!

**Reply by einarpersson** [reply]
i'm sorry, you are right
